README.rst
=================
Invoice Generator
=================


.. image:: https://img.shields.io/pypi/v/french-invoice-generator.svg
        :target: https://pypi.python.org/pypi/french-invoice-generator

.. image:: https://img.shields.io/travis/tgrandjean/french-invoice-generator.svg
        :target: https://app.travis-ci.com/github/tgrandjean/french-invoice-generator

.. image:: https://readthedocs.org/projects/french-invoice-generator/badge/?version=latest
        :target: https://invoice-generator.readthedocs.io/en/latest/?version=latest
        :alt: Documentation Status




generate invoices with latex from python


* Free software: Apache Software License 2.0
* Documentation: https://invoice-generator.readthedocs.io.


Features
--------

* TODO

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.0 (2021-10-20)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for invoice_generator."""


tests/test_invoice_generator.py
#!/usr/bin/env python

"""Tests for `invoice_generator` package."""
from datetime import date
import os
from pathlib import Path
import pytest
from jinja2 import Template
from invoice_generator import invoice_generator, models


@pytest.fixture
def prestation():
    return models.Prestation(title="prestation",
                             unit_price=10.0,
                             quantity=2,
                             vat=10)


@pytest.fixture
def invoice(prestation):
    data = {
        "reference": "2021-001",
        "emited": date.today(),
        "issuer": models.Issuer(
            **{"company_name": "My Awesome Company with characters_to_escape!",
               "first_name": "Pierre",
               "last_name": "Qui roule",
               "siret": '0000000000',
               "intracom_vat": 'FRXXXXXXXXX',
               "address": models.Address(address="Champs de Mars",
                                         zip_code=75000,
                                         city='Paris'),
               "email": "contact@my-awesome-company.com",
               "rib": models.RIB(name="my awesome company",
                                 bic="FR XXX",
                                 iban="FR 00000 0000 0000")
               }
        ),
        "customer": models.Customer(
            first_name='Jean',
            last_name="Dupond",
            name="Dupond's company",
            email="jean.dupond@example.com",
            address=models.Address(address="17 Avenue des Champs-Elysées",
                                   zip_code=75000,
                                   city="Paris"),
            phone="+33 6 00 00 00 00"
        ),
        "prestations": [
            prestation
        ]
    }
    return models.Invoice(**data)


@pytest.fixture
def invoice_escaped(invoice):
    escaped_company_name = r"My Awesome Company with characters\_to\_escape!"
    invoice.issuer.company_name = escaped_company_name
    return invoice


@pytest.fixture
def project_path():
    return Path(__file__).resolve().parents[1]


@pytest.fixture
def template_dir(project_path):
    return project_path / 'invoice_generator' / 'templates'


@pytest.fixture
def generator(invoice, template_dir, tmpdir):
    return invoice_generator.InvoiceGenerator(invoice,
                                              template_dir,
                                              'main.tex',
                                              tmpdir,
                                              'test')


def test_prestation_total_validator(prestation):
    assert prestation.total == 20


def test_prestation_total_vat_validator(prestation):
    assert prestation.total_vat == 2


def test_invoice_total_without_charge_validator(invoice):
    assert invoice.total_without_charge == 20


def test_invoice_total_vat_validator(invoice):
    assert invoice.total_vat == 2


def test_invoice_total_validator(invoice):
    assert invoice.total == 22


def test_invoice_pagination(invoice):
    invoice.prestations = invoice.prestations * 11
    assert len(invoice.paginated_prestations) == 2
    assert len(invoice.paginated_prestations[0]) == 10
    invoice.prestations = invoice.prestations * 2
    assert len(invoice.paginated_prestations) == 2
    assert len(invoice.paginated_prestations[0]) == 16
    invoice.prestations = invoice.prestations * 2
    assert len(invoice.paginated_prestations) == 3
    assert len(invoice.paginated_prestations[0]) == 16
    assert len(invoice.paginated_prestations[1]) == 20
    assert len(invoice.paginated_prestations[2]) == 8


def test_invoice_generator_template_dir_accessor(generator, template_dir):
    assert generator.template_dir == template_dir


def test_invoice_generator_bad_template_dir(generator, project_path):
    with pytest.raises(FileNotFoundError):
        generator.template_dir = project_path / "bad_directory"


def test_invoice_generator_template_dir_conversion(generator, template_dir):
    generator.template_dir = str(template_dir)
    assert generator.template_dir == template_dir


def test_invoice_generator_template_dir_setter(generator, template_dir):
    generator.template_dir = None
    assert generator.template_dir == template_dir


def test_invoice_generator_template_name_accessor(generator):
    assert generator.template_name == 'main.tex'


def test_invoice_generator_template_name_setter(generator):
    generator.template_name = None
    assert generator.template_name == "main.tex"


def test_invoice_generator_invoice_name_accessor(generator):
    assert generator.invoice_name == 'test'


def test_invoice_generator_invoice_name_setter(generator):
    generator.invoice_name = 'test.pdf'
    assert generator.invoice_name == 'test'


def test_invoice_generator_data_accessor(generator, invoice_escaped):
    assert generator.data == invoice_escaped


def test_invoice_generator_output_directory_setter(generator, template_dir):
    generator.output_directory = None
    assert generator.output_directory == template_dir


def test_invoice_generator_output_directory_setter_str(generator,
                                                       template_dir):
    generator.output_directory = str(template_dir)
    assert generator.output_directory == template_dir


def test_invoice_generator_file_to_compile_accessor(generator, tmpdir):
    assert generator._file_to_compile == tmpdir / 'test'


def test_invoice_generator_load_template(generator):
    generator._load_template()
    assert isinstance(generator._template, Template)


def test_invoice_generator_run(generator, tmpdir):
    generator.run()
    assert os.listdir(tmpdir) == ['test.pdf']


def test_invoice_generator_clean(generator, tmpdir):
    generator._load_template()
    generator._generate_tex()
    generator._compile_latex()._compile_latex()
    assert len(os.listdir(tmpdir)) == 6
    generator._clean()
    assert len(os.listdir(tmpdir)) == 1


def test_invoice_generator_compilation_failed(generator):
    with pytest.raises(ValueError):
        generator._compile_latex()


generate_examples.py
from invoice_generator import InvoiceGenerator
from invoice_generator.models import (Address, Customer, Issuer,
                                      Invoice, Prestation)
from datetime import datetime
from random import randint, random


def example(n_prestations, invoice_name):
    issuer = Issuer(
        first_name="first_name",
        last_name="last_name",
        company_name="company_name",
        siret="siret",
        intracom_vat="intracom_vat",
        email="email@example.com",
        phone="phone number",
        address=Address(address='address', zip_code=75000, city="Paris")
    )
    customer = Customer(
        first_name="first_name",
        last_name="last_name",
        # name="company_name",
        address=Address(address="address", zip_code=75000, city="Paris"),
        phone="phone number",
        email="email@example.com"
    )
    prestations = [
        Prestation(quantity=randint(1, 10),
                   unit_price=random() * 10,
                   title=f"prestation {i+1}"
                   ) for i in range(n_prestations)
    ]
    invoice = Invoice(issuer=issuer,
                      customer=customer,
                      prestations=prestations,
                      reference="reference",
                      emited=datetime.now())
    InvoiceGenerator(invoice,
                     invoice_name=invoice_name,
                     output_directory="./example").run()


if __name__ == "__main__":
    example(1, 'example_1')
    example(8, 'example_2')
    example(26, 'example_3')


invoice_generator/models.py
"""Pydantic models for invoice_generator."""
from datetime import date
from typing import Optional, List
from pydantic import BaseModel, EmailStr, validator


class Address(BaseModel):
    address: str
    zip_code: int
    city: str


class RIB(BaseModel):
    name: Optional[str]
    iban: str
    bic: str


class Customer(BaseModel):
    name: Optional[str]
    first_name: Optional[str]
    last_name: Optional[str]
    address: Optional[Address]
    email: Optional[EmailStr]
    phone: Optional[str]


class Prestation(BaseModel):
    title: str
    unit_price: float
    quantity: float
    vat: float = 0.0
    total: float = None
    total_vat: float = None

    @validator('total', always=True, pre=True)
    def compute_total(cls, v, values):
        return values['quantity'] * values['unit_price']

    @validator('total_vat', always=True, pre=True)
    def compute_total_vat(cls, v, values):
        return values["unit_price"] * values['quantity'] \
            * (values['vat'] / 100)


class Issuer(BaseModel):
    first_name: str
    last_name: str
    company_name: str
    address: Optional[Address]
    siret: str
    intracom_vat: str
    logo: Optional[str]
    rib: Optional[RIB]
    email: EmailStr
    phone: Optional[str]


class Invoice(BaseModel):
    title: str = 'Facture'
    reference: str
    emited:  date
    issuer: Issuer
    customer: Customer
    prestations: List[Prestation]
    total_without_charge: float = None
    total_vat: float = None
    total: float = None
    payment_within: int = 30
    late_payment_message: Optional[str]

    @validator('total_without_charge', always=True, pre=True)
    def compute_total_without_charge(cls, v, values):
        return sum([presta.total for presta in values["prestations"]])

    @validator('total_vat', always=True, pre=True)
    def compute_total_vat(cls, v, values):
        return sum([presta.total_vat for presta in values["prestations"]])

    @validator('total', always=True, pre=True)
    def compute_total(cls, v, values):
        return values['total_without_charge'] + values['total_vat']

    @property
    def total_by_vat(self):
        _ = dict.fromkeys(set([presta.vat for presta in self.prestations]),
                          0.0)
        for presta in self.prestations:
            _[presta.vat] += presta.total_vat
        return _

    @property
    def paginated_prestations(self):
        if len(self.prestations) <= 8:
            return [self.prestations]
        elif len(self.prestations) <= 12:
            return [self.prestations[:10], self.prestations[10:]]
        elif len(self.prestations) <= 24:
            return [self.prestations[:16], self.prestations[16:]]
        else:
            pages = []
            pages.append(self.prestations[:16])  # first page
            values = self.prestations[16:]
            num_pages = len(values) // 20
            for i in range(num_pages):
                pages.append(values[i * 20: (i + 1) * 20])
            offset = num_pages * 20
            pages.append(values[offset:])  # last page
            return pages


invoice_generator/invoice_generator.py
"""Main module."""
import os
from pathlib import Path
import re
import subprocess
import uuid
import jinja2

from .models import Invoice


JINJA_CONF = {
    "block_start_string": r'\BLOCK{',
    "block_end_string": '}',
    "variable_start_string": r'\VAR{',
    "variable_end_string": '}',
    "comment_start_string": r'\#{',
    "comment_end_string": '}',
    "line_statement_prefix": '%%',
    "line_comment_prefix": '%#',
    "trim_blocks": True,
    "autoescape": False,
}


class InvoiceGenerator:
    """Invoice Generator.

    Generate an invoice in pdf using LaTeX.

    :param template_dir: Filepath to the directory that contains LaTeX
        templates.
    :type template_dir: pathlib.Path or str
    :param template_name: The name of that LaTeX template to use.
    :type template_name: str
    :param data: The data to use for populating the template.
    :type data: Invoice
    :param output_directory: Path of output directory
    :type output_directory: pathlib.Path or str
    :param invoice_name: The name of the outputed pdf, defaults to None. If
        no invoice_name are passed to generator, it will generate one using
        uuid4.
    :type invoice_name: str, optional
    """

    def __init__(self,
                 data,
                 template_dir=None,
                 template_name=None,
                 output_directory=None,
                 invoice_name=None):
        self.template_dir = template_dir
        self.template_name = template_name
        self.invoice_name = invoice_name
        self.output_directory = output_directory
        self.data = data

    @property
    def template_dir(self):
        return self._template_dir

    @template_dir.setter
    def template_dir(self, template_dir):
        if not template_dir:
            template_dir = Path(__file__).resolve().parents[0] / 'templates'
        if not os.path.exists(template_dir):
            msg = f"The directory {template_dir} doens't exists."
            raise FileNotFoundError(msg)
        if type(template_dir) == str:
            template_dir = Path(template_dir).resolve()
        self._template_dir = template_dir

    @property
    def output_directory(self):
        return self._output_directory

    @output_directory.setter
    def output_directory(self, output_directory):
        if not output_directory:
            output_directory = self.template_dir
        if type(output_directory) == str:
            output_directory = Path(output_directory).resolve()
        self._output_directory = output_directory

    @property
    def template_name(self):
        return self._template_name

    @template_name.setter
    def template_name(self, template_name):
        if not template_name:
            template_name = "main.tex"
        self._template_name = template_name

    @property
    def invoice_name(self):
        return self._invoice_name

    @invoice_name.setter
    def invoice_name(self, invoice_name):
        invoice_name = invoice_name or str(uuid.uuid4())
        if invoice_name.endswith('.pdf'):
            self._invoice_name.replace('.pdf', '')
        else:
            self._invoice_name = invoice_name

    @property
    def data(self):
        return self._data

    @data.setter
    def data(self, data):
        data = self._escape_latex_characters(data.dict())
        self._data = Invoice(**data)

    @property
    def _file_to_compile(self):
        return self.output_directory / (self.invoice_name)

    @property
    def _latex_jinja_env(self):
        loader = jinja2.FileSystemLoader(self.template_dir)
        return jinja2.Environment(loader=loader, **JINJA_CONF)

    @staticmethod
    def _tex_escape(text):
        """
        :param text: a plain text message
        :return: the message escaped to appear correctly in LaTeX

        from https://stackoverflow.com/questions/16259923/how-can-i-escape-latex-special-characters-inside-django-templates  # noqa
        """
        conv = {
            '&': r'\&',
            '%': r'\%',
            '$': r'\$',
            '#': r'\#',
            '_': r'\_',
            '{': r'\{',
            '}': r'\}',
            '~': r'\textasciitilde{}',
            '^': r'\^{}',
            '\\': r'\textbackslash{}',
            '<': r'\textless{}',
            '>': r'\textgreater{}',
        }
        regex = re.compile('|'.join(re.escape(str(key)) for key in
                           sorted(conv.keys(), key=lambda item: - len(item))))
        return regex.sub(lambda match: conv[match.group()], text)

    @staticmethod
    def _escape_latex_characters(data):
        """
        Recursively escape all values in a dictionary
        :param data:
        :return: escaped dictionary
        :rtype: dict

        Adapted from https://gist.github.com/zenweasel/8bf8b4dfed2c5d2c8805
        """
        if isinstance(data, list):
            for x, l in enumerate(data):
                if isinstance(l, dict) or isinstance(l, list):
                    InvoiceGenerator._escape_latex_characters(l)
                else:
                    if type(l) == str:
                        data[x] = InvoiceGenerator._tex_escape(l)

        if isinstance(data, dict):
            for k, v in data.items():
                if isinstance(v, dict) or isinstance(v, list):
                    InvoiceGenerator._escape_latex_characters(v)
                else:
                    if type(v) == str:
                        data[k] = InvoiceGenerator._tex_escape(v)
            return data

    def _load_template(self):
        self._template = self._latex_jinja_env\
                                .get_template(self.template_name)

    def _generate_tex(self):
        to_compile = self._template.render(invoice=self._data)
        with open(str(self._file_to_compile) + '.tex', 'w') as file:
            file.write(to_compile)

    def _check_compilation_success(self):
        if not re.search('Output written on', self.__stdout.decode()):
            raise ValueError('Compilation failed')

    def _clean(self):
        for dirname, _, filenames in os.walk(self.output_directory):
            for f_name in filenames:
                if f_name.startswith(self._invoice_name):
                    if not f_name.endswith('.pdf'):
                        os.remove(os.path.join(dirname, f_name))

    def _compile_latex(self):
        cmd = ['pdflatex',
               "-synctex=1",
               "-interaction=nonstopmode",
               '-output-directory',
               self.output_directory,
               self._file_to_compile
               ]
        process = subprocess.Popen(cmd,
                                   cwd=self._template_dir,
                                   stdout=subprocess.PIPE,
                                   stderr=subprocess.PIPE)
        self.__stdout, self.__stderr = process.communicate()
        self._check_compilation_success()

        return self

    def run(self, clean=True):
        self._load_template()
        self._generate_tex()
        self._compile_latex()._compile_latex()
        if clean:
            self._clean()
        return self.output_directory / (self.invoice_name + '.pdf')


mock.py
from invoice_generator.invoice_generator import InvoiceGenerator
from invoice_generator.models import (Address, Invoice, Issuer, RIB,
                                      Customer, Prestation)
from datetime import date
from random import randint, random, choice


def fake_prestations(number, variable_vat=False):
    return [
        Prestation(title=f"prestation {i + 1}",
                   unit_price=random() * randint(1, 10),
                   quantity=randint(1, 10),
                   vat=choice([10, 20]) if variable_vat else 0
                   ) for i in range(number)
    ]


data = {
    "reference": "2021-001",
    "emited": date.today(),
    "issuer": Issuer(
        **{"company_name": "My Awesome Company",
           "first_name": "Pierre",
           "last_name": "Qui roule",
           "siret": '0000000000',
           "intracom_vat": 'FRXXXXXXXXX',
           "address": Address(address="Champs de Mars",
                              zip_code=75000,
                              city='Paris'),
           "email": "contact@my-awesome-company.com",
           "rib": RIB(name="my awesome company",
                      bic="FR XXX",
                      iban="FR 00000 0000 0000")
           }
    ),
    "customer": Customer(
        first_name='Jean',
        last_name="Dupond",
        name="SARL Dupond et fils",
        email="jean.dupond@example.com",
        address=Address(address="17 Avenue des Champs-Elysées",
                        zip_code=75000,
                        city="Paris"),
        phone="+33 6 00 00 00 00"
    ),
    "prestations": fake_prestations(1, variable_vat=False)
}
data = Invoice(**data)

invoice_generator = InvoiceGenerator('./templates',
                                     'main.tex',
                                     data,
                                     output_directory='./example',
                                     invoice_name='test_2').run()
